[V1][TPU] Support V1 Sampler for ragged attention

NickLucche opened the pull request:
Updated version of https://github.com/vllm-project/vllm/pull/13982 with ragged kernel attention.

All considerations described in previous PR attempt are still valid (in particular still aiming to get a ~~single~~ optimized graph with model+sampling), the main difference is in how the number of sampled tokens should be handled now that it's not always equal to the batch size.

I think there's fundamentally one trade-off to consider:
 - A: we sample on all the tokens in the unified batch, even those that do not need it. This way the input to the sampler is fixed so we don't recompile. Num tokens that need sampling range from 1..`max_num_seqs`.
 - B:  we select only the logits  that require sampling before `model.sample` and eat the cost of compilation (as it happens on main now).

I think there's also a third option which I tried to implement here:
 - C: we pre-compile a bunch of fixed-sized tensors, and we only feed those sizes to the Sampler. It's basically the logits thatneed sampling are selected + padding to the nearest pre-compiled shape. This means the input to the sampler and consequently the output of the ModelWrapper is of some pre-compiled fixed shape. Eg. `do_sample_indices: [4, 10]=>padded_do_sample_indices: [4, 10, 0, 0]`

Note: ~compilation test is currently failing (re-running a request with different sampling params isn't as fast), will dig more into it.~ Execution just isn't as fast as it was in previous PR: compiling doesn't *always* drop times by >10x. This is happening on main too, so I don't think it has anything to do with this PR. As a result, I lowered the bar in `test_sampler_compilation.py`. We can focus on optimizing performance later (this is anyway the first PR where we try to track it).  

Tested with `VLLM_USE_V1=1 python -m pytest -s tests/v1/tpu/test_sampler.py`.

### Update:
I broke up again the model + sampling graph, so now compilation stages are separated:
```sh
INFO 03-06 08:57:12 [tpu_model_runner.py:908] Compiling the model with different input shapes.
INFO 03-06 08:57:12 [tpu_model_runner.py:913]   -- num_tokens: 16
INFO 03-06 08:57:51 [tpu_model_runner.py:913]   -- num_tokens: 32
INFO 03-06 08:58:53 [tpu_model_runner.py:913]   -- num_tokens: 64
INFO 03-06 09:01:12 [tpu_model_runner.py:921] Compilation finished in in 240.17 [secs].
INFO 03-06 09:01:12 [tpu_model_runner.py:923] Compiling sampling with different input shapes.
INFO 03-06 09:01:12 [tpu_model_runner.py:937]   -- num_tokens: 16, num_seqs: 8
INFO 03-06 09:01:12 [tpu_model_runner.py:937]   -- num_tokens: 16, num_seqs: 16
INFO 03-06 09:01:12 [tpu_model_runner.py:937]   -- num_tokens: 16, num_seqs: 32
INFO 03-06 09:01:12 [tpu_model_runner.py:937]   -- num_tokens: 16, num_seqs: 64
INFO 03-06 09:01:12 [tpu_model_runner.py:937]   -- num_tokens: 32, num_seqs: 64
INFO 03-06 09:01:12 [tpu_model_runner.py:937]   -- num_tokens: 64, num_seqs: 64
INFO 03-06 09:01:13 [tpu_model_runner.py:949] Compilation finished in in 0.39 [secs].
```

NickLucche committed:
xla friendly minp+topk

Signed-off-by: NickLucche <nlucches@redhat.com>

NickLucche committed:
fix platform check

Signed-off-by: NickLucche <nlucches@redhat.com>

NickLucche committed:
tracing sampler

Signed-off-by: NickLucche <nlucches@redhat.com>

NickLucche committed:
forward_tpu + revert topk selection

Signed-off-by: NickLucche <nlucches@redhat.com>

NickLucche committed:
refactor to avoid recompiling None values and disable topk/topp

Signed-off-by: NickLucche <nlucches@redhat.com>

NickLucche committed:
tests + updated torch dev version

Signed-off-by: NickLucche <nlucches@redhat.com>

NickLucche committed:
wip: adapt to ragged attn kernel

Signed-off-by: NickLucche <nlucches@redhat.com>

NickLucche committed:
adapt to ragged attn kernel and multimodal

Signed-off-by: NickLucche <nlucches@redhat.com>

NickLucche committed:
add tests

Signed-off-by: NickLucche <nlucches@redhat.com>

NickLucche commented:
Also, sadly compilation takes forever now. Waiting for new kernels to address that.

mgoin commented on vllm/v1/worker/tpu_model_runner.py:
```diff
@@ -744,12 +910,20 @@ def capture_model(self) -> None:
-            self._dummy_run(self.kv_caches, num_tokens)
-            logger.info("  -- num_tokens: %d", num_tokens)
-            xm.mark_step()
-            xm.wait_device_ops()
-            if num_tokens >= self.max_num_tokens:
+            # Compile for different sampler outputs in
+            # [MIN_NUM_SEQS, pad(max_num_reqs)]
+            while True:
+                self._dummy_run(self.kv_caches, num_tokens)
+                logger.info("  -- num_tokens: %d, num_seqs: %d", num_tokens,
+                            num_reqs_to_sample)
+                xm.mark_step()
+                xm.wait_device_ops()
+                if num_reqs_to_sample >= self.max_num_reqs:
+                    break
+                num_reqs_to_sample *= 2
+            if num_tokens >= self.scheduler_config.max_num_batched_tokens:
                 break
             num_tokens *= 2
```
This doesn't seem right.. each time we are only passing in `num_tokens`, so what is the point of `num_reqs_to_sample` in this loop? Looking around it seems like `num_tokens_to_sample` is missing to be passed in.

This potential explosion of compilation makes me feel a better solution would be to completely separate the graphs for model forward and sampling. This way we can just define a "max sampling batch size" that can be separated - so we end up with M+N compilations rather than M*N

NickLucche replied:
I don't think I can do it in in M+N here as sampling will be dependent on both `num_reqs_to_sample` and *`hidden_state`*, so for each different hidden_state shape we still need to compile. But we will only have MN compilations of the sampling step which should be much lighter.
 
I did consider breaking the graph multiple times but iirc the main concern was the interaction with torch.compile.

mgoin commented on vllm/v1/worker/tpu_model_runner.py:
```diff
@@ -849,12 +1020,40 @@ def forward(
+            kv_caches: The key and value caches. They can be None during the
+                memory profiling at initialization.
+        """
+        # NxV, N num padded tokens
```
What is V here?

mgoin commented on vllm/v1/worker/tpu_model_runner.py:
```diff
@@ -673,16 +837,17 @@ def load_model(self) -> None:
         xm.mark_step()
         xm.wait_device_ops()
         model = ModelWrapperV1(model)
+        self.model = model
```
cruft?

NickLucche replied:
yeah rebase leftover

mgoin commented on vllm/v1/worker/tpu_model_runner.py:
```diff
@@ -728,14 +893,15 @@ def _dummy_run(
-                kv_caches=kv_caches,
-                inputs_embeds=inputs_embeds,
-            )
+            self.model(input_ids, position_ids, kv_caches, sampling_meta)
```
Please keep the named args here and add back `inputs_embeds`

NickLucche committed:
break up model|sample graph to speed up compilation

Signed-off-by: NickLucche <nlucches@redhat.com>

NickLucche committed:
minor check on optional temp

Signed-off-by: NickLucche <nlucches@redhat.com>

NickLucche commented:
Quickly looking into an issue with greedy sampling and the control flow introduced here https://github.com/vllm-project/vllm/pull/13587

NickLucche committed:
fix greedy sampling

Signed-off-by: NickLucche <nlucches@redhat.com>

alexm-redhat reviewed (commented):
@NickLucche Nice work with the sampler and the optimizations to improve compilations times. Left some comments and questions.

NickLucche committed:
move tpu sampling params in own file

Signed-off-by: NickLucche <nlucches@redhat.com>

NickLucche commented on vllm/v1/worker/tpu_model_runner.py:
```diff
@@ -745,13 +920,41 @@ def capture_model(self) -> None:
             logger.info("  -- num_tokens: %d", num_tokens)
+            self._dummy_run(self.kv_caches, num_tokens)
             xm.mark_step()
-            xm.wait_device_ops()
```
🙏🏻🙏🏻

NickLucche commented on vllm/v1/sample/sampler.py:
```diff
@@ -210,8 +210,8 @@ def apply_min_p(
-        # Apply mask using boolean indexing
-        logits[~valid_token_mask] = -float('inf')
+        # Apply mask using boolean indexing (xla friendly)
+        logits.masked_fill_(~valid_token_mask, -float("inf"))
```
I think I was inspecting the compilations with `PT_XLA_DEBUG=2`

robertgshaw2-redhat replied:
Should we revert this or does it also help GPu performance?

NickLucche replied:
good point, I can add this in a separate PR if needs be now that the two samplers are separated.

NickLucche commented on tests/v1/tpu/test_sampler.py:
```diff
@@ -0,0 +1,97 @@
+        s = time()
+        _ = llm.generate(prompts, sampling_params)
+        run2 = time() - s
+        assert run1 > run2
```
performance is back with the kernel update in https://github.com/vllm-project/vllm/pull/14310, so I'll bring back the 0.1 check :)

NickLucche commented on tests/v1/tpu/test_sampler.py:
```diff
@@ -0,0 +1,97 @@
+    # compiled for in previous eager iteration.
+    sampling_params = SamplingParams(temperature=0.1, min_p=0.8, max_tokens=64)
+    output2 = llm.generate(prompts, sampling_params)
+    assert output[0].outputs[0].text != output2[0].outputs[0].text
```
done

NickLucche commented on vllm/v1/worker/tpu_model_runner.py:
```diff
@@ -42,6 +45,160 @@
+    frequency_penalties = None
+    presence_penalties = None
+    repetition_penalties = None
+    # should use tensor
```
yep but they're still disabled for now, I'd like to add args incrementally following a deeper analysis of performances

NickLucche commented on vllm/v1/worker/tpu_model_runner.py:
```diff
@@ -42,6 +45,160 @@
+            do_argmax=None) -> "TPUSupportedSamplingMetadata":
+        # As sampling happens on a single traced function, options
+        # are "disabled" by having them evaluate to an Identity op.
+        # Note that initialization is dependent on num_samples.
```
I can elaborate a bit in the code too, what I mean is that we trace `Sampler.forward` by turning on all supported params (min_p and temp here) so we don't have to compile one graph for each flow in Sampler.forward (basically every "if" triggers re-compilation unless we use tensors).
In following runs we can "disable" min_p by setting its value to 0.0 to that it does not trigger recompile and achieves same output.

NickLucche commented on vllm/v1/worker/tpu_model_runner.py:
```diff
@@ -745,13 +920,41 @@ def capture_model(self) -> None:
+            if num_tokens >= self.scheduler_config.max_num_batched_tokens:
+                break
+            num_tokens *= 2
+        xm.wait_device_ops()
```
see comment below :)

NickLucche replied:
unfortunately it doesn't make much of a difference when compiling the graph, not sure if there's some env var we can set.
It does make a lot of difference obv when executing the compiled graph though.

NickLucche committed:
address review

Signed-off-by: NickLucche <nlucches@redhat.com>

NickLucche commented:
Thanks a lot for the review and comments @alexm-redhat !
Just rebased on top of https://github.com/vllm-project/vllm/pull/14310 and performances are back to normal. I also addressed your comments!

NickLucche committed:
rebase cruft

Signed-off-by: NickLucche <nlucches@redhat.com>

mgoin reviewed (approved):
Awesome work building the structure for future sampling param enablement once we get performant kernels!

Eval smoke test looks good (and runs fast!)
```
Running generate_until requests: 100%|██████████████████████| 1319/1319 [01:44<00:00, 12.67it/s]
INFO:lm-eval:Output path not provided, skipping saving results aggregated
vllm (pretrained=Qwen/Qwen2-1.5B-Instruct,max_model_len=2048,trust_remote_code=True), gen_kwargs: (None), limit: None, num_fewshot: 5, batch_size: auto
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.5868|±  |0.0136|
|     |       |strict-match    |     5|exact_match|↑  |0.5815|±  |0.0136|
```

mgoin commented on vllm/v1/worker/tpu_model_runner.py:
```diff
@@ -746,13 +768,41 @@ def capture_model(self) -> None:
             xm.mark_step()
-            xm.wait_device_ops()
-            if num_tokens >= self.max_num_tokens:
+            if num_tokens >= self.scheduler_config.max_num_batched_tokens:
```
Actually I think we should use `self.max_num_tokens` still here since I made this be the padded number of tokens, in case the user put in an odd number or something weird

mgoin commented on vllm/v1/worker/tpu_model_runner.py:
```diff
@@ -746,13 +768,41 @@ def capture_model(self) -> None:
+                if num_reqs_to_sample >= self.max_num_reqs:
+                    break
+                num_reqs_to_sample *= 2
+            if num_tokens >= self.scheduler_config.max_num_batched_tokens:
```
ditto

NickLucche committed:
max_num_tokens stopping condition when compiling

Signed-off-by: NickLucche <nlucches@redhat.com>

yaochengji commented on vllm/v1/worker/tpu_model_runner.py:
```diff
@@ -576,6 +589,20 @@ def execute_model(
+        # mismatch in pre-processing, it will trigger a small recompilation
+        # of the code thus far. Forward graph remains untouched.
+        tpu_sampling_metadata = TPUSupportedSamplingMetadata.\
+            from_sampling_metadata(sampling_metadata, padded_do_sample_indices,
```
There's recompilation after warmup here.

NickLucche replied:
can you share the logs to help me debug this? It's not showing on the tests we have, perhaps we can add more to prevent it from happening. I will double check with `enfore_eager=False` which is not tested rn.

NickLucche replied:
nvm see comment below.

yaochengji commented on vllm/v1/worker/tpu_model_runner.py:
```diff
@@ -620,7 +644,11 @@ def execute_model(
         if max_gen_len == 1:
+            # ModelWrapper returns reqs output + padding. Remove padding
+            # here to keep dynamic op outside of main graph.
             valid_sampled_token_ids = selected_token_ids.tolist()
```
There's recompilation after warmup here.

NickLucche replied:
In the `tolist()` move? How come?

yaochengji replied:
Because `tolist` moves data from TPU to CPU, and as `selected_token_ids`'s value is not ready, a graph capture and recompilation will be trigged.

NickLucche replied:
oh sure, I thought it would just call mark_step on forced sync and realize post processing graph didn't change after the first inference

yaochengji reviewed (commented):
Hi @NickLucche , thanks for your contribution, I found two places causing recompilation. Is is possible to resolve them?

WoosukKwon reviewed (changes requested):
Sorry for chiming in this late, but I think we need a broader discussion on how to share the sampler between different hardware backends. In V0, we found that the sampler implementation got over-complicated partly because different hardware backends added some `if` statements here and there just for their own purposes. I do really want to prevent it from happening this time.

mgoin commented:
I think we are fine to duplicate the sampler, and we will likely need to once we get to integrating custom kernels. This was just a minimal implementation.

NickLucche commented:
Thanks for reviwing @yaochengji ! 

The  re-compilation you're referring to is actually the one I mention in this TODO https://github.com/vllm-project/vllm/pull/14227/files#diff-e2942ece30a5c580437694ffb964bfc664b510c59244c08e5921b8f5cefb4280R560.
If we add a mark_step there we can see it happening. `from_sampling_metadata` just happens to be the only mark_step exec we call after then.
I was planning to address that in a follow up PR tbh, but since the PR is stalling I might get more time to work on this.
This is most likely caused by the creation of the actual `SamplingMetatadata` obj inside `self.input_batch` *the first time* it is called. 
I didn't want to drag this too long as given the current state of TPU code I have to constantly rebase.

yaochengji commented:
@NickLucche Thanks for your contribution!

I'm thinking about whether we can put the computation in `torch.compile` if possible. It brings two benefits:
1. the compilation boundary is much more clean compared with `xm.mark_step`
2. the graph trace overhead is lower.

NickLucche committed:
rebase changes

Signed-off-by: NickLucche <nlucches@redhat.com>

NickLucche committed:
fix capture_graph loop

Signed-off-by: NickLucche <nlucches@redhat.com>

NickLucche committed:
fix recompilation issue on sampling graph; add new tpu sampler

Signed-off-by: NickLucche <nlucches@redhat.com>

NickLucche committed:
newline conflict(?)

Signed-off-by: NickLucche <nlucches@redhat.com>

NickLucche commented:
Rebased and updated.

@WoosukKwon I've separated the Sampler code for TPU into a different namespace so it doesn't make use of the CUDA path any longer. Let me know what you think.

@yaochengji I've addressed the small recompilation issue that was happening due to a mismatch in pre-processing at compile time vs runtime.
Please note that there's still a small recompilation happening on 2nd runs which I pinpointed to `vllm.v1.worker.gpu_input_batch.InputBatch._make_sampling_metadata` slicing tensors to comply with V1 "data structures reuse policy" . 
If that's ok, I'd like to address that in a separate PR as this one has gotten quite clunky already.

yaochengji commented:
@NickLucche Thanks for your update!

Previously when resolving the recompilation issue in logits processor, I noticed the compilation time is about 500ms. This is not acceptable in some use cases. (E.g. a typical requirement of TPOT is 100ms). Therefore, could you resolve all the recompilation issues?

BTW, there's an ongoing task to ensure that there's no recompilation after warmup, https://github.com/vllm-project/vllm/issues/14580.

NickLucche commented:
This is a different recompilation happening in pre rather than post processing, and it is dependent on existing v1 code.

alexm-redhat reviewed (approved):
LGTM!

NickLucche committed:
Merge branch 'main' into tpu-sampler-ragged

NickLucche committed:
revert gpu sampler change

Signed-off-by: NickLucche <nlucches@redhat.com>

hyeygit commented:
Thank you @NickLucche for this PR! Do you plan to merge it soon? I hope to send a follow-up PR to optimize top-k for TPU.

NickLucche commented:
@hyeygit Thanks for your optimization work, feel free to ping me when you have something ready
